src/periphery/MarketLens.sol

```diff
@@ -0,0 +1,76 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+// Uncomment this line to use console.log
+
+import "interfaces/IBentoBoxV1.sol";
+import "interfaces/ICauldronV2.sol";
+import {IERC20, IStrictERC20} from "BoringSolidity/interfaces/IERC20.sol";
+import "utils/CauldronLib.sol";
+
+contract MarketLens {
+    address payable public owner;
+    uint8 constant DECIMALS = 18;
+
+    constructor() {
+        owner = payable(msg.sender);
+    }
+
+    function getInterestPerYear(ICauldronV2 cauldron) external view returns (uint64 interestPerYear) {
+        (, , uint64 interestPerSecond) = cauldron.accrueInfo();
+        interestPerYear = CauldronLib.getInterestPerYearFromInterestPerSecond(interestPerSecond);
+    }
+
+    function getMaxBorrowForCauldronV2(ICauldronV2 cauldron) external view returns (uint256 maxBorrow) {
+        IBentoBoxV1 bentoBox = IBentoBoxV1(cauldron.bentoBox());
+        IERC20 mim = IERC20(cauldron.magicInternetMoney());
+        uint256 poolBalance = bentoBox.balanceOf(mim, address(cauldron));
+        maxBorrow = bentoBox.toAmount(mim, poolBalance, false);
+    }
+
+    function getMaxBorrowForCauldronV3(ICauldronV3 cauldron) external returns (uint256 maxBorrow) {
+        IBentoBoxV1 bentoBox = IBentoBoxV1(cauldron.bentoBox());
+        IERC20 mim = IERC20(cauldron.magicInternetMoney());
+        uint256 poolBalance = bentoBox.balanceOf(mim, address(cauldron));
+        uint256 mimInBentoBox = bentoBox.toAmount(mim, poolBalance, false);
+        uint256 userBorrowLimit = getUserBorrowLimit(cauldron);
+        maxBorrow = min(userBorrowLimit, mimInBentoBox);
+    }
+
+    // Why does this modify state? Want to be able to use view but cannot.
+    function getUserBorrowLimit(ICauldronV3 cauldron) public returns (uint256 userBorrowLimit) {
+        (uint256 totalLimit, uint256 borrowPartPerAddress) = cauldron.borrowLimit();
+        userBorrowLimit = min(totalLimit, borrowPartPerAddress);
+    }
+
+    function getTotalMimBorrowed(ICauldronV2 cauldron) external view returns (uint256 totalMimBorrowed) {
+        Rebase memory totalBorrow = cauldron.totalBorrow();
+        totalMimBorrowed = totalBorrow.elastic;
+    }
+
+    function getTvl(ICauldronV2 cauldron) external view returns (uint256 tvl) {
+        IBentoBoxV1 bentoBox = IBentoBoxV1(cauldron.bentoBox());
+        uint256 totalTokensDeposited = bentoBox.toAmount(cauldron.collateral(), cauldron.totalCollateralShare(), false);
+        tvl = (totalTokensDeposited * 10 ** DECIMALS) / getOracleExchangeRate(cauldron);
```

Comment:
Not sure what the best way to do this would be... I wanted to originally get the cauldron.collateral() and then get .decimals(), but it was complaining. I saw there's some pattern of using Boring's IStrictERC20 to get decimals, but not sure if I should do that...

Reply:
this is good as the getOracleExchangeRate is going to report the price (inversed) in the same decimal precision as the collateral. Current WBTC oracle reports 4284,  so totalCollateralShare is 25693071566 so 25693071566 / 4284 = 5997449. Multiplying it by 1e18 is more precise and totally fine, it's just that you need to / 1e18 and consider it scaled.

I wouldn't say DECIMALS in this case but more like 

```
tvl = (totalTokensDeposited * 1e18) / getOracleExchangeRate(cauldron);
```

scaling by 1e18 is pretty common... you can also make a constant named TVL_PRECISION = 1e18, it's up to you, it's probably cleaner.

Reply:
Makes sense!